src/main/java/com/conveyal/r5/analyst/Grid.java

```diff
@@ -154,9 +159,27 @@ public void rasterize (Geometry geometry, double value) {
                 Geometry pixel = getPixelGeometry(x + west, y + north, zoom);
                 Geometry intersection = pixel.intersection(geometry);
                 double weight = intersection.getArea() / area;
-                grid[x][y] += weight * value;
+                weights.put(new int[] { x, y }, weight);
             }
         }
+
+        return weights;
+    }
+
+    /**
+     * Do pycnoplactic mapping:
+     * the value associated with the supplied polygon a polygon will be split out proportionately to
+     * all the web Mercator pixels that intersect it.
+     */
+    public void rasterize (Geometry geometry, double value) {
+        incrementFromPixelWeights(getPixelWeights(geometry), value);
+    }
+
+    public void incrementFromPixelWeights (TObjectDoubleMap<int[]> weights, double value) {
+        for (TObjectDoubleIterator<int[]> it = weights.iterator(); it.hasNext();) {
+            it.advance();
+            grid[it.key()[0]][it.key()[1]] += it.value() * value;
```

Comment:
Wow, this really makes me wish Java had built in tuple types. I guess as long as there's no major performance hit, it's sane to use a nice structured type like a Map from int[2] to double to pass around the pixel weights. Java seems pretty good at creating and destroying a lot of objects quickly. But `grid[it.key()[0]][it.key()[1]]` is still not really that readable. What about a list of (x, y, weight)? It requires a new class definition but might be clearer what's going on.